### PR TITLE
fix: do not crash PutObjectTags when a node is down

### LINKED ISSUE #10939: mc tag set returns 502 if not all nodes are online in distributed setup
## Expected Behavior

I expected tagging to be available when I lose one node in distributed setup.

## Current Behavior

mc tag set returns 502 when one node goes offline in my 12 nodes cluster. There is a LB in front of minio cluster. 

image

If I access one of the nodes directly, the error is different

```
# mc tag set staas-test-direct/test/test.txt "key1=test"                                                              │·············································
mc: <ERROR> Failed to set tags for http://node.example.com:9000/test/test.txt. Put "http://node.example.com:9000│·············································
/test/test.txt?tagging=": Connection closed by foreign host http://node.example.com:9000/test/test.txt?tagging=. Ret│·············································
ry again.
```

## Steps to Reproduce (for bugs)

1. Deploy minio in distributed mode
2. Shutdown one node
3. Try to set a tag on any object

## Context

We can't shutdown even one node for maintenance in our production distributed setup without affecting users who use bucket tagging.

## Your Environment

* Version used (`minio --version`): RELEASE.2020-10-28T08-16-50Z and RELEASE.2020-05-01T22-19-14Z
* Operating System and version (`uname -a`): CentOS 7

### PR DESCRIPTION
**Pull request by harshavardhana**
## Description
fix: do not crash PutObjectTags when a node is down

## Motivation and Context
fixes #10939

## How to test this PR?
As per #10939 

## Types of changes
- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [ ] Documentation needed
- [ ] Unit tests needed

### CODE REVIEW

**Comment by minio-trusted** [comment]
## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-large-bucket.sh | :heavy_check_mark: |
| mint-fs.sh | :heavy_check_mark: |
| mint-gateway-s3.sh | :heavy_check_mark: |
| mint-erasure.sh | :heavy_check_mark: |
| mint-dist-erasure.sh | :heavy_check_mark: |
| mint-zoned.sh | :heavy_check_mark: |
| mint-gateway-nas.sh | :heavy_check_mark: |
| mint-gateway-azure.sh | :x: [more...](#3dba8c92-892e-44b5-bfbc-3bf73fcb312f) |
<a name="3dba8c92-892e-44b5-bfbc-3bf73fcb312f"></a>
### 10940-ba4ef3e/mint-gateway-azure.sh.log:
```
Running with
SERVER_ENDPOINT:      minio-dev8.minio.io:30394
ACCESS_KEY:           minioazure
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0

To get logs, run 'docker cp 7a910ca0c87a:/mint/log /tmp/mint-logs'

(1/15) Running aws-sdk-go tests ... done in 8 seconds
(2/15) Running aws-sdk-java tests ... done in 2 seconds
(3/15) Running aws-sdk-php tests ... done in 2 minutes and 27 seconds
(4/15) Running aws-sdk-ruby tests ... done in 21 seconds
(5/15) Running awscli tests ... done in 2 minutes and 48 seconds
(6/15) Running healthcheck tests ... done in 1 seconds
(7/15) Running mc tests ... done in 3 minutes and 23 seconds
(8/15) Running minio-dotnet tests ... done in 1 minutes and 36 seconds
(9/15) Running minio-go tests ... FAILED in 1 minutes and 49 seconds
{
  "args": {
    "bucketName": "minio-go-test-nxq3lp1zeex6u5sb",
    "objectName": "test-object",
    "opts": "",
    "size": -1
  },
  "duration": 1178,
  "function": "PutObject(bucketName, objectName, reader,size,opts)",
  "message": "Unexpected size",
  "name": "minio-go: testPutObjectStreaming",
  "status": "FAIL"
}
(9/15) Running minio-java tests ... done in 10 minutes and 16 seconds
(10/15) Running minio-js tests ... FAILED in 47 seconds
{
  "name": "minio-js",
  "function": "\"after all\" hook in \"functional tests\"",
  "duration": 80,
  "status": "FAIL",
  "error": "S3Error: The bucket you tried to delete is not empty at Object.parseError (node_modules/minio/dist/main/xml-parsers.js:86:11) at /mint/run/core/minio-js/node_modules/minio/dist/main/transformers.js:156:22 at DestroyableTransform._flush (node_modules/minio/dist/main/transformers.js:80:10) at DestroyableTransform.prefinish (node_modules/readable-stream/lib/_stream_transform.js:129:10) at prefinish (node_modules/readable-stream/lib/_stream_writable.js:611:14) at finishMaybe (node_modules/readable-stream/lib/_stream_writable.js:620:5) at endWritable (node_modules/readable-stream/lib/_stream_writable.js:643:3) at DestroyableTransform.Writable.end (node_modules/readable-stream/lib/_stream_writable.js:571:22) at IncomingMessage.onend (_stream_readable.js:682:10) at endReadableNT (_stream_readable.js:1252:12) at processTicksAndRejections (internal/process/task_queues.js:80:21)"
}
(10/15) Running minio-py tests ... done in 18 minutes and 37 seconds
(11/15) Running s3cmd tests ... done in 2 minutes and 12 seconds
(12/15) Running s3select tests ... done in 37 seconds
(13/15) Running security tests ... done in 0 seconds

Executed 13 out of 15 tests successfully.
```
Deleting image on docker hub
Deleting image locally

**Review by poornas — APPROVED** [review]
LGTM

File: cmd/erasure-object.go [hunk 1]
```diff
@@ -1071,32 +1075,27 @@ func (er erasureObjects) PutObjectTags(ctx context.Context, bucket, object strin
...
-		if tags != "" {
-			fi.Metadata[xhttp.AmzObjectTagging] = tags
-		}
-		for k, v := range opts.UserDefined {
-			fi.Metadata[k] = v
+	onlineDisks, metaArr = shuffleDisksAndPartsMetadataByIndex(onlineDisks, metaArr, fi.Erasure.Distribution)
+	for i, metaFi := range metaArr {
+		if metaFi.IsValid() {
+			// clean fi.Meta of tag key, before updating the new tags
+			delete(metaFi.Metadata, xhttp.AmzObjectTagging)
```
Comment (vadmeste) [inline cmd/erasure-object.go:1089]: This is from the old code but doesn't seem to be needed, shouldn't we only update if tags is not empty ?

**Review by vadmeste — COMMENTED** [review]
One comment

File: cmd/erasure-object.go (hunk 1 above)
Comment (harshavardhana) [reply cmd/erasure-object.go:1089]: There is no update all tags a replace only operations @vadmeste - delete is a no-op if key doesn't exist so its okay.

**Review by vadmeste — APPROVED** [review]
LGTM

**Pull request merged**